hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/DatanodeDescriptor.java

```diff
@@ -1007,6 +1013,7 @@ public void updateRegInfo(DatanodeID nodeReg) {
     for(DatanodeStorageInfo storage : getStorageInfos()) {
       if (storage.getStorageType() != StorageType.PROVIDED) {
         storage.setBlockReportCount(0);
+        storage.setBlockContentsStale(true);
```

Comment:
Why set content stale here?

Reply:
Thanks @zhangshuyan0 for you comment.

The modifications here may not be directly related to the current problem.
the reason for modifying this is that if the current dn is re-registered, block deletion or exception may occur on the dn during this period, because FBR has not yet been completed, and the NN side memory record is different from the actual dn block.

If processExtraRedundancyBlock is executed at this time, block loss may occur. 
such as a file has 2 replicas, but only 1 replica is expected.  when processExtraRedundancyBlock is executed, a live dn will be choose for deletion, which will cause a block miss, so the dn re-registering needs to be marked blockContentsStale will avoid this case.

Reply:
Thank @haiyang1987  for your reply. I understand what you mean. This patch  removes corresponding excess replicas from ExcessRedundancyMap when re-registering, so NameNode does not know whether the replica is still on the register DN any more.

Reply:
Thanks @haiyang1987 for your report. And thanks @zhangshuyan0 for your review.

I think this modification is a new bug, not related to this case, we need to fix this bug in a new issue. 

We can produce this bug by the following steps:

- Assume that there is block1 contains three replicas, dn1, dn2
- DN1 is shutdown for maintenance for corrupt disk
- Admin removed the corrupted disk and restart datanode 
- DN1 try to register it to NameNode through registerDatanode rpc
- End-user try to decrease the replicas of block from 2 to 1 through setReplication RPC
- Block1 still contains three replicas in namenode, but the dn1 is not existed because it is stored in a corrupt disk
- NameNode select dn2 as a redundancy replica for this block to delete
- DN1 try to report all stored blocks to namnode through blockreport rpc
- NameNode will remove the dn1 replica for block1 because the blockreport from DN1 doesn't contains block1

After these two operations(setReplication from end-user and restart from admin), the block1 may lose all replicas.
So I think we should mark all storage as a stale storage while namenode processing registerdatanode rpc, so that this case can be fixed.

@zhangshuyan0 @haiyang1987 I'm looking forward your good idea, thanks.

Reply:
@ZanderXu Thanks for your reply. I think this modification is not a new bug. Before this patch, NameNode knows all excess replicas even though a DataNode is re-registered, so it wouldn't delete more replicas than expected. 
As the situation you just said, we can discuss it from two aspects:
1. If NameNode knows the corrupt replicas corresponding to corrupt disk, it will not delete the only healthy replica.
2. If NameNode know nothing about the corrupt disk, the essence of the problem is that the Admin manually removed some replicas without notifying NameNode. Then in the time between "the replica has been removed" and "NN learns that the replica has been removed", there is always a chance that the only healthy replica will be deleted. So, I think the key to solving this problem is to immediately notify NN which disk is corrupt. Restarting & re-registering may not be necessary.

Reply:
@zhangshuyan0 Thanks for your reply.  This case has nothing to do with excess replicas. The mainly problem is the time between "registerDataNode" and "blockReport". 

During in this time, namenode thinks that this DN contains this block, but actually the DN doesn't store this block(let's not discuss how this block was lost in this DN first). So during this time, namenode shouldn't delete any replicas for this block, right? NameNode can only delete replicas of this block after "blockReport", right?

Reply:
HDFS-16999 that you did is a very nice catch. It fix a problem that namenode may contains the incorrect metadata. 

This case I said above is used to prevent namenode to delete any replicas depended on uncertain or stale metadata in NameNode.

Reply:
> namenode thinks that this DN contains this block, but actually the DN doesn't store this block

Actually, this situation can happen at any time, not just between "registerDataNode" and "blockReport". Why do you think that after the DN is re-registered, the probability of the above situation happening will increase, and it needs to be dealt with specifically?

IMO, the concept stale was born to mark replicas that may have been deleted by NN commands but do not exist in the ExcessRedundancyMap. It is closely related to the state of ExcessRedundancyMap. 
- After the failover, there is no data in the ExcessRedundancyMap of the new ANN, so all datanodes need to be marked as stale. 
- In this patch, the ExcessRedundancyMap of the corresponding DN is cleared when re-registered, so it needs to be set to stale.

If the data in the ExcessRedundancyMap is correct, it means that NN clearly knows which replicas are about to be deleted, which ensures that NN will not actively delete all replicas of a block.

Looking forward to your reply @ZanderXu .

Reply:
> Actually, this situation can happen at any time, not just between "registerDataNode" and "blockReport". Why do you think that after the DN is re-registered, the probability of the above situation happening will increase, and it needs to be dealt with specifically?

Yes, there are some other situations can cause  this case. And I don't think the probability of the above situation happening will increase. 

I just think we have a chance to reduce this case. So I think we need to do it.

Reply:
About the relationship between stale storage and ExcessRedundancyMap. 

-  StaleStorage is used to prevent the namenode from deleting replicas of blocks whose replicas are indeterminate.
-  ExcessRedundancyMap is used to mark the replicas of blocks that namenode is deleting

Reply:
@ZanderXu I think there is a misunderstanding between us. I totally agree with this change.  The difference between us may be that I think it is more appropriate to merge this change with this PR instead of opening a new issue.

Reply:
> StaleStorage is used to prevent the namenode from deleting replicas of blocks whose replicas are indeterminate.

About StaleStorage we can say comments in code:
https://github.com/apache/hadoop/blob/42e695d510befabf3b7cfc52349b0d78faadb249/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/DatanodeStorageInfo.java#L142-L149

> At startup or at failover, the storages in the cluster may have pending block deletions from a previous incarnation of the NameNode.

From this, it can be seen that, the design of the "stale content" is to address the "indeterminate" caused by pending deletions. By the way, if the information provided by ExcessRedundancyMap is accurate, there will be no "indeterminate" caused by pending deletions.
See also: https://issues.apache.org/jira/browse/HDFS-1972

Reply:
Sorry, late reply.
Thanks @ZanderXu @zhangshuyan0 for your detailed reply. It is very meaningful to me and  learned a lot.